numpy/core/src/multiarray/compiled_base.c

```diff
@@ -500,6 +506,37 @@ binary_search(double key, double arr [], npy_intp len)
     if (key > arr[len - 1]) {
         return len;
     }
+    else if (key < arr[0]) {
+        return -1;
+    }
+
+    if (guess < 0)  {
+        guess = 0;
+    }
+    else if (guess >= len - 1) {
+        guess = len - 2;
```

Comment:
With `len == 1` or `len == 2` you can go into this branch with `guess == 0` or `guess == 1`, which will leave `guess` pointing outside the array, and cause a segmentation fault in the next few lines.

Not sure what the best solution is. I think the ideal situation would be to special-case anything with `len < 4` (I would suggest a linear search ignoring `guess`), and for the larger arrays make sure in this normalization that `guess >= 1` and `guess <= len - 3`, so that the testing of the three most likely intervals can be done with no extra checking with a couple of nested branches, e.g.:

```
assert(guess >=1 && guess <= len-3);

if (key > arr[guess]) {
    if (key <= arr[guess + 1]) {
        return guess;
    }
    else if (key <= arr[guess + 2]) {
        return guess + 1;
    }
    imin = guess + 2;
}
else if (key > arr[guess - 1]) {
    return guess - 1;
}
else {
    imax = guess - 1;
}
```